### PR TITLE
Fix span tagging for application error in GO YARPC/gRPC

### PR DESCRIPTION
**Pull request by kexiongliu123**
Fix span tagging for application error in GO YARPC/gRPC
- In https://github.com/yarpc/yarpc-go/pull/2282/files we captured invokeErr
- In addition to the invokeErr, in this PR we also emit application errors to span tags
- The error message is intentionally omitted to prevent exposing personally identifiable information (PII) in tracing systems.

### CODE REVIEW

File: api/transport/propagation.go [hunk 1]
```diff
@@ -22,6 +22,7 @@ package transport
 
 import (
 	"context"
+	"go.uber.org/yarpc/yarpcerrors"
```
Comment (biosvs) [inline api/transport/propagation.go:25]: Please sort imports properly

File: api/transport/propagation.go [hunk 2]
```diff
@@ -37,6 +38,11 @@ type CreateOpenTracingSpan struct {
 	ExtraTags     opentracing.Tags
 }
 
+const (
+	//TracingTagStatusCode is the span tag key for the YAPRC status code.
```
Comment (biosvs) [inline api/transport/propagation.go:42]: Add space plz

File: api/transport/propagation.go [hunk 3]
```diff
@@ -119,3 +125,17 @@ func UpdateSpanWithErr(span opentracing.Span, err error) error {
 	}
 	return err
 }
+
+// UpdateSpanWithoutErrMsg sets the error tag and status code on a span.
+// The error message is intentionally omitted to prevent exposing
+// personally identifiable information (PII) in tracing systems.
+// Returns the given error
```
Comment (biosvs) [inline api/transport/propagation.go:132]: Let's indeed add error returning to the function signature and body.

File: api/transport/propagation.go [hunk 4]
```diff
@@ -119,3 +125,17 @@ func UpdateSpanWithErr(span opentracing.Span, err error) error {
...
+func UpdateSpanWithoutErrMsg(span opentracing.Span, err error, errCode yarpcerrors.Code) {
+	if err != nil {
+		span.SetTag("error", true)
```
Comment (biosvs) [inline api/transport/propagation.go:135]: Should it has any special tags like "application error" or something?

File: internal/observability/call.go [hunk 5]
```diff
@@ -496,6 +499,17 @@ func (c call) logStreamEvent(err error, success bool, succMsg, errMsg string, ex
 	ce.Write(fields...)
 }
 
+// emitSpanErrorTag sets the error information as tags on the current span
+func (c call) emitSpanErrorTag(res callResult) {
+	if span := opentracing.SpanFromContext(c.ctx); span != nil {
+		// emit application error to span tag if applicable
+		if res.isApplicationError {
```
Comment (biosvs) [inline internal/observability/call.go:506]: Let's check this boolean first (if !.. {return}), it's much cheaper than creating span from context.

File: internal/observability/call.go [hunk 6]
```diff
@@ -496,6 +499,17 @@ func (c call) logStreamEvent(err error, success bool, succMsg, errMsg string, ex
 	ce.Write(fields...)
 }
 
+// emitSpanErrorTag sets the error information as tags on the current span
+func (c call) emitSpanErrorTag(res callResult) {
```
Comment (biosvs) [inline internal/observability/call.go:503]: Let's indicate in the name that this tags should be set only for application error.

File: internal/observability/call.go [hunk 7]
```diff
@@ -496,6 +499,17 @@ func (c call) logStreamEvent(err error, success bool, succMsg, errMsg string, ex
...
+	if span := opentracing.SpanFromContext(c.ctx); span != nil {
+		// emit application error to span tag if applicable
+		if res.isApplicationError {
+			transport.UpdateSpanWithoutErrMsg(span,
+				res.err, yarpcerrors.FromError(res.err).Code())
```
Comment (biosvs) [inline internal/observability/call.go:508]: Could it be that error will be set here already? In other PRs you're setting errors, including message, without checking if it's an application error.

Should we replace all other PRs with changes inside observability/call?

**Review by biosvs — CHANGES_REQUESTED** [review]

File: internal/observability/call.go [hunk 8]
```diff
@@ -496,6 +499,17 @@ func (c call) logStreamEvent(err error, success bool, succMsg, errMsg string, ex
 	ce.Write(fields...)
 }
 
+// emitSpanErrorTag sets the error information as tags on the current span
+func (c call) emitSpanErrorTag(res callResult) {
+	if span := opentracing.SpanFromContext(c.ctx); span != nil {
```
Comment (ChenX1993) [inline internal/observability/call.go:504]: My concern of using obs middleware is that the required inbound/ outbound spans are not available in go context when you are trying to obtain spans here.

Could you double check?

Comment (kexiongliu123) [reply internal/observability/call.go:508]: After reviewing the code more thoroughly, I believe you're right. It doesn’t seem straightforward to pass the original spans here. Do you have any suggestions on how we might approach this?

**Review by biosvs — CHANGES_REQUESTED** [review]

**Pull request closed without merging**